src/test/java/org/elasticsearch/NamingConventionTests.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch;
+
+import org.elasticsearch.test.ElasticsearchTestCase;
+
+import java.io.IOException;
+import java.lang.reflect.Modifier;
+import java.net.URISyntaxException;
+import java.nio.file.*;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.regex.Pattern;
+
+/**
+ * Simple class that ensures that all subclasses concrete of ElasticsearchTestCase end with either Test | Tests
+ */
+public class NamingConventionTests extends ElasticsearchTestCase {
+
+    // see https://github.com/elasticsearch/elasticsearch/issues/9945
+    public void testNamingConventions()
+            throws ClassNotFoundException, IOException, URISyntaxException {
+        final List<Class> classes = new ArrayList<>();
+        String[] packages = {"org.elasticsearch", "org.apache.lucene"};
+        for (final String packageName : packages) {
+            final String path = "/" + packageName.replace('.', '/');
+            Path startPath = Paths.get(NamingConventionTests.class.getResource(path).toURI());
+
+            Files.walkFileTree(startPath, new FileVisitor<Path>() {
+                private Path pkgPrefix = Paths.get(path).getParent();
+                @Override
+                public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs) throws IOException {
+                    pkgPrefix = pkgPrefix.resolve(dir.getFileName());
+                    return FileVisitResult.CONTINUE;
+                }
+
+                @Override
+                public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
+                    String filename = file.getFileName().toString();
+                    if (filename.endsWith("Tests.class") || filename.endsWith("Test.class")) { // don't worry about the ones that match the pattern
+                        return FileVisitResult.CONTINUE;
```

Comment:
not directly related, but shall we also check that all of these do extend `ElasticsearchTestCase`?